### PR TITLE
Add regression test that reproduces inconsistent state bug

### PR DESCRIPTION
**Pull request by ruuda**
We observed a bug in production that caused us to hit this case:

https://github.com/channable/hoff/blob/483d68fefdf30f9076e1899b7143bd85ccce176e/src/Logic.hs#L467

This pull request:

* Extends the test interpreter so we can pass lists of results per operation, so we can build more complex scenario's in tests.
* Adds a test that hits that error case.

The test currently fails, the next step will be making it pass, but because this is quite a large diff already, I wanted to open a separate pull request for the test.

### CODE REVIEW

File: tests/Spec.hs [hunk 1]
```diff
@@ -565,6 +621,55 @@ main = hspec $ do
...
+        results = defaultResults
+          { resultIntegrate =
+            [ Right $ Sha "b71"
+            , Left $ Logic.IntegrationFailure (Branch "master")
+            ]
+          , resultPush =
+            [ PushRejected
+              -- TODO(ruuda): Why is this second result needed? We should not do
+              -- a second push.
+            , PushOk
```
Comment (rightfold) [inline tests/Spec.hs:656]: Will this be fixed together with the inconsistent state bug?

Comment (ruuda) [reply tests/Spec.hs:656]: Yes, once it no longer hits the error case there, I hope to be able to further debug this test.

**Comment by ruuda** [comment]
This will be merged together with its fix in #59.

File: tests/Spec.hs (hunk 1 above)
Comment (ruuda) [reply tests/Spec.hs:656]: This has also been addressed in #59.

**Pull request closed without merging**